### ISSUE TITLE
"mx-run ... ClassName" makes ps output useless

### ISSUE THREAD

**Issue by preaction** [moose/MooseX-Runnable#3]
https://github.com/moose/MooseX-Runnable/blob/master/script/mx-run#L26

I believe I currently work at the place MXR was built for, so I'm sure there was some reason for that line, but at the moment, it is causing me more pain than anything: It doesn't tell me which Perl I ran it with, it doesn't tell me which perl flags I ran it with, it doesn't tell me which arguments I ran it with.

Would it be terrible if I removed that line?

**Comment by stevan** [comment]
I would think that someone somewhere is relying on that, why not make it optional? The default should be to do what it does now, but you could add an option to override/turn-off the behavior, should be fairly simple I would think.

**Comment by karenetheridge** [comment]
hey wow, people are still using MooseX::Runnable? :)

**Comment by preaction** [comment]
Yeah, we're even building new things with it (and [wrote a presentation about it](http://chicago.pm.org/Command-Line-Apps/#49)). With the amount of ad-hoc runnable things we build, it's a lot easier to organize modules in $PERL5LIB than modulinos on $PATH (though it's also hard to get people to move to new practices).

Perhaps I could do, for backcompat purposes:

```
local $0 = 'mx-run ... ' . $class_name . ' ' . $0;
```

But only because I can't think of any good way to configure not altering $0: A command-line flag is easily forgotten, a class attribute nice but strange, and both potentially trampling something that exists.

Another possibility:

```
package My::Runnable;
use Moose;
with qw( MooseX::Runnable );

our $MX_RUNNABLE_PRESERVE_PROGNAME = 1;
```

But that's kind of ... no.

I'm also just, in general, opposed to changing $0 except to hide sensitive information from the process table.

**Comment by shadowcat-mst** [comment]
On Thu, Jan 29, 2015 at 04:47:07PM -0800, Doug Bell wrote:
> I'm also just, in general, opposed to changing $0 except to hide sensitive information from the process table.

My guess is that it was implemented for one of two things: (1) 'hide passwords
in @ARGV from ps', (2) 'make mx-run instances for a particular class easier
to find in ps'.

I'd find the former sort of reasonable and the second grounds for handing
our sysadmin a cluebat and coming back after the screaming stopped.

-- 
Matt S Trout - Shadowcat Systems - Perl consulting with a commit bit and a clue

http://shadowcat.co.uk/blog/matt-s-trout/   http://twitter.com/shadowcat_mst/

Email me now on mst (at) shadowcat.co.uk and let's chat about how our CPAN
commercial support, training and consultancy packages could help your team.
